### PR TITLE
Implement a calendar algorithm. Passes all unit tests.

### PR DESCRIPTION
**Pull request by saraholiviawu**
**Background**: Implement a calendar algorithm that, given meeting information, returns all available times for which a requested meeting can occur that day. The calendar also supports  optional attendees, such that if one or more time slots exist for both mandatory and optional attendees to attend, those slots will be returned. Otherwise, the time slots that fit just mandatory attendees are returned.

**Testing**: This algorithm passes all 21 unit test provided and 5 more tests I have written.

### CODE REVIEW

**Review by smruthibalajee — APPROVED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,208 @@
...
+    // If only optional attendees are requested, return available time ranges for optional attendees.
+    if (attendees.size() == 0) {
+      ArrayList<TimeRange> availableTimeRanges = returnAvailableTimeRanges(events, optionalAttendees, requestDuration);
+      return availableTimeRanges;
+    }
+
+    // Else: Assume that both mandatory and optional attendees are requested.
+    ArrayList<TimeRange> mandatoryAvailableTimeRanges = returnAvailableTimeRanges(events, attendees, requestDuration);
+    ArrayList<TimeRange> optionalAvailableTimeRanges = returnAvailableTimeRanges(events, optionalAttendees, requestDuration);
+    ArrayList<TimeRange> mutualAvailableTimeRanges = getMutualAvailableTimeRanges(mandatoryAvailableTimeRanges, optionalAvailableTimeRanges, requestDuration);
```
Comment (peter14f) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: it's good practice to keep no more than 80 characters in one line

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,208 @@
...
+    }
+    // Merge any overlapping time ranges
+    ArrayList<TimeRange> mergedBusyTimeRanges = getMergedBusyTimeRanges(sortedBusyTimeRanges);
+    // Get available time ranges
+    ArrayList<TimeRange> availableTimeRanges = getAvailableTimeRanges(mergedBusyTimeRanges, requestDuration);
+    return availableTimeRanges;
+  }
+
+  // Returns overlapping time ranges between mandatory & optional time ranges that have a valid duration.
+  private ArrayList<TimeRange> getMutualAvailableTimeRanges(ArrayList<TimeRange> mandatoryAvailableTimeRanges, ArrayList<TimeRange> optionalAvailableTimeRanges, int requestDuration) {
```
Comment (peter14f) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: ditto

**Review by peter14f — APPROVED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (saraholiviawu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Fixed

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (saraholiviawu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Fixed

**Pull request merged**